modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,253 @@
+        this.placementDriver = placementDriver;
+        this.catalogService = catalogService;
+        this.clock = clock;
+    }
+
+    /**
+     * Gets primary replicas by partition for the table.
+     *
+     * @param tableId Table ID.
+     * @return Primary replicas for the table, or null when not yet known.
+     */
+    public CompletableFuture<List<ReplicaHolder>> primaryReplicasAsync(int tableId) {
+        return primaryReplicas.compute(tableId, (id, hld) -> {
+            if (hld == null || hld.replicas.isCompletedExceptionally()) {
+                return new Holder(initReplicasForTableAsync(id));
+            }
+
+            hld.lastAccessTime = System.currentTimeMillis();
+            return hld;
+        }).replicas;
+    }
+
+    long updateCount() {
+        return updateCount.get();
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void start() {
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        catalogService.listen(CatalogEvent.TABLE_DROP, (EventListener) this);
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void stop() {
+        catalogService.removeListener(CatalogEvent.TABLE_DROP, (EventListener) this);
+        placementDriver.removeListener(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        primaryReplicas.clear();
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> initReplicasForTableAsync(Integer tableId) {
+        try {
+            // Initially, request all primary replicas for the table.
+            // Then keep them updated via PRIMARY_REPLICA_ELECTED events.
+            long timestamp = clock.nowLong();
+            CatalogTableDescriptor tableDesc = catalogService.table(tableId, timestamp);
+
+            if (tableDesc == null) {
+                // TODO: Should we wait for table sync?
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            CatalogZoneDescriptor zoneDesc = catalogService.zone(tableDesc.zoneId(), timestamp);
+
+            if (zoneDesc == null) {
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            return primaryReplicasAsyncInternal(tableId, zoneDesc.partitions());
+        } catch (Throwable t) {
+            return CompletableFuture.failedFuture(t);
+        }
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> primaryReplicasAsyncInternal(int tableId, int partitions) {
+        CompletableFuture<ReplicaMeta>[] futs = (CompletableFuture<ReplicaMeta>[]) new CompletableFuture[partitions];
+
+        for (int partition = 0; partition < partitions; partition++) {
+            futs[partition] = placementDriver.getPrimaryReplica(new TablePartitionId(tableId, partition), clock.now());
+        }
+
+        CompletableFuture<Void> all = CompletableFuture.allOf(futs);
+
+        return all.thenApply(v -> {
```

Comment:
I'd rather add an exception handler with a corresponding assertion.

Reply:
Could you please elaborate? Not sure I understand.

Reply:
Let's say that one of futs was completed exceptionally in that case `all` will also fail and thenApply won't be called and we will basically miss map update. We either not expect futs to fail or should add corresponding exception handler. In any case silent no-op will complicate bug analysis.

Reply:
Good catch. Fixed.